Graphics/MusicWheelItem Song NormalPart/get_lamp.lua

```diff
@@ -0,0 +1,86 @@
+local args = {...}
+local graphic = args[1]
+local pn = args[2]
+
+LampColors = {
+    [0]={1,1,1,0},
+    color "#ffffff",
+    color "#34bfff",
+    color "#48ff1d",
+    color "#fff568",
+    color "#bfeaff"
+}
+
+
+local function getLamp(self, song)
+    local best_lamp = 0
+
+    if song then
+        -- Default to the machine profile if not found
+        if PROFILEMAN:IsPersistentProfile(pn) then 
+          profile = PROFILEMAN:GetProfile(pn)
+        else 
+          profile = PROFILEMAN:GetMachineProfile()
+        end
+                
+        local diff = GAMESTATE:GetCurrentSteps(pn):GetDifficulty()
+        local st = GAMESTATE:GetCurrentStyle():GetStepsType()
+        local steps = song:GetOneSteps(st, diff)
+        if steps then
+            high_score_list = profile:GetHighScoreListIfExists(song, steps)
+            
+            -- Unplayed song (or played, but quit out)
+            if high_score_list == nil then
+                return 0
+            end
+            
+            for score in ivalues(high_score_list:GetHighScores()) do
+                if score then
```

Comment:
Can we generally just use the `StageAward` for each of the highescore and find the best one:
https://itgmania.github.io/lua-for-itgmania/LuaAPI#Classes-HighScore-GetStageAward

That should work for ITG mode and then only for FA+ mode we can might need to compute if we have a great full combo (as there's no `FullComboW4`) if necessary.

Reply:
Question before I replace it, what's the verdict on "Good FCs", do we even want to bother with a lamp for those? Here we're getting into DDR vs. ITG philosophy for what counts as an FC or not.  I was leaning toward DDR-like lamps for anything that doesn't have a miss in it, but I could go either way.

Another thought, what's the current behavior if you don't FC with respect to the `StageAward`? Does it just not populate in the Stats.xml? Checking my own, I see even failed songs are showing `FullComboW3`.

And just to be clear, W4 represents "Good" under ITG mode and "Great" under FA+? I'm a bit less familiar with FA+ mode.

Reply:
Because SL is predominantly an ITG theme we should cater to that. I do intend on adding a DDR mode at some point but it's still a bit away. So for now we should always just be using lamps up to Great Full Combo. This would mean in ITG mode we have:
- Full Fantastic Combo (W1)
- Full Excellent Combo (W2)
- Full Great Combo (W3)

While in FA+ mode we have:
- Full Fantastic+ Combo (W1)
- Full Fantastic Combo (W2)
- Full Excellent Combo (W3)
- Full Great Combo (needs to be computed manually as there isn't a corresponding `StageAward` for `FullComboW4` for it).

Also when checking the `StageAwards` you can just filter out the failed songs based on the grade. In general we should always try to use engine functions if they're available, and it seems like `StageAward` is essentially doing what you want.

(Fwiw I don't quite understand the question of "what's the current behavior if you don't FC with respect to the StageAward")

Reply:
I'm not so sure we can rely on the `StageAward`, take a look at this entry in my personal stats.xml,

```
<HighScore>
<Name>BWAG</Name>
<Grade>Tier07</Grade>
<Score>93869026</Score>
<PercentDP>0.903700</PercentDP>
<SurviveSeconds>113.488480</SurviveSeconds>
<MaxCombo>233</MaxCombo>
<StageAward>FullComboW3</StageAward>
<PeakComboAward>1000</PeakComboAward>
...
<ProductID>1</ProductID>
<TapNoteScores>
<HitMine>0</HitMine>
<AvoidMine>0</AvoidMine>
<CheckpointMiss>0</CheckpointMiss>
<Miss>5</Miss>
<W5>0</W5>
<W4>1</W4>
<W3>16</W3>
<W2>83</W2>
<W1>334</W1>
<ProW5>0</ProW5>
<ProW4>0</ProW4>
<ProW3>0</ProW3>
<ProW2>0</ProW2>
<ProW1>0</ProW1>
<MaxScore>0</MaxScore>
<CheckpointHit>0</CheckpointHit>
</TapNoteScores>
<HoldNoteScores>
<LetGo>2</LetGo>
<Held>39</Held>
<MissedHold>0</MissedHold>
</HoldNoteScores>
...
</HighScore>
```

This is in ITG mode, and you can see W4 is present (and a few misses), but the StageAward shown is FullComboW3--why is that the case?

https://itgmania.github.io/lua-for-itgmania/LuaAPI#Enums-StageAward

I would expect there to be no stage award (or at least some enum value to represent no award). Is there something I'm missing?

Reply:
Assuming this behavior is expected, it means we can't use the FullComboW3 award, so for ITG mode we'd need to manually calculate Greats, and FA+ mode both Greats and Excellents. If DDR mode were to be added, then goods would need tracking as well, bringing the total to 3 unaccounted for awards.

I would like to use the awards to simplify the logic, but based on the options today, it seems like it would only work in about half the cases anyway. Let me know if I'm off base with this at all.

Reply:
Can you post a link to that stats.xml? I'm wondering what's happening here. I haven't seen the `ProW5-W1` entries before either.

Reply:
Sure,

https://github.com/bwaggone/Simply-B/blob/stats/stats/Stats.xml

FYI this is on a machine running windows + outfox. Mostly the August 1st build.

Reply:
Hmm I'm not sure what's happening here. I looked at various Stats.xml's I have (running SL with ITGmania as well as on StepMania 5.1b2) and the StageAward seems to be populated correctly while it seems to be completely wrong in yours. There are 0 instances of `<StageAward></StageAward>` (indicating no stage awards) which is very suspect. Every `<HighScore>` of yours has a `<StageAward>` populated which is incorrect.

Here's mine for example: https://gist.github.com/teejusb/9cc1a1abd21172d671e5f6e8a6ebdb05

Reply:
Hmm... Could be a quirk from outfox, my OS, or library versions I suppose? My stats file has some other weird quirks I've noticed with the way it keeps generating new hashes for existing songs... I can try a few different installs and see if there's any patterns. What OS are you running your versions of stepmania on?

In either case, considering FA+ mode (and the eventual DDR mode) will need a special case to handle FullComboW4, maybe we keep this implementation? Other themes seem to use similar implementations; and depending on how consistent my stageaward bug is across systems it might be better to skip past it with this implementation as well.

Reply:
I generally play on Linux and test on Windows although I do have a mac install as well. I generally run ITGmania but have a StepMania 5.1b2 install that I originally ported the stats.xml from.

I generally do strongly prefer leveraging engine resources as much as possible and then using lua to fill in the gaps as needed. DDR us still quite a whiles away so it's only needing to support one case for FA+ mode.

Reply:
Looks to be an Outfox specific bug, created an issue for their team,

https://github.com/TeamRizu/OutFox/issues/642

I totally agree with you--would rather not duplicate work that's already given to us for free. Given that the stageaward behavior is a bug--would you rather I forgo Outfox support? Or leave a TODO to clean this up once they fix it on their end?

fwiw I do recall seeing an outfox fork of SimplyLove--so we could leave it to the downstream consumers there to ensure their own compatibility.

Reply:
I think we should use the engine things as much as possible, even here, so I would opt to use StageAward in this PR. Iirc here is this SL fork here: https://github.com/Tiny-Foxes/Simply-Love-OutFox . While we do list Outfox in the `SupportedVersions` I think downstream bugs of engines shouldn't be something we want to implement workarounds for, especially since both Outfox and ITGmania are under active development (exception being 5.0.12/5.1b2, since those aren't under active dev).

Reply:
Is using stage awards the best idea? Iirc, the saved stage award is based exclusively off of the highest scoring play and not based on the best achieved stage award ever on a particular song. This would mean that a FullComboW3 could easily get overwritten by a higher scoring run with a miss, and you would end up losing the lamp. Is that the desired behavior?

Reply:
There's a `StageAward` line for each `HighScore` that is saved and not just one per song, same as the judgment breakdowns for each of the `HighScores` which is what this PR currently looks at. You are correct in that higher scoring with worse `StageAwards` can get overwritten, but that is still the case either way because that's just how the engine handles writing scores to stats.xml in general.

The engine *should* in theory always save a score with the best `StageAward` but currently it does not. ITGmania does intend to add that functionality but I don't have a timeline yet.
 
If one wants StageAwards to be completely decoupled from the score, at the moment one WILL need to write a separate file outside of the engine to do all of the score tracking (like what Waterfall does, which SL does not intend to).

Reply:
There appears to be a few more wrinkles with this approach--FA+ mode doesn't seem to save ANY StageAwards. Tested both in ITGMania and Outfox, the former just leaves them empty no matter the proper award and the latter still erroneously awards FullComboW3 in every case. Am I missing something here?

Additionally, score:GetStageAward() returns a string rather than an enum as I would have thought. I can create a common table to convert it to an int or just raw compare against strings, but that feels a bit like reinventing the enum that should already be present. Is there a way to convert the result of score:GetStageAward() to that enum easily?

Given both of those hiccups, what's your recommendation? Or am I overcomplicating things?

Reply:
There are no enums in Lua. Every enum also has a reverse map which should be accessible via `<EnumName>:Reverse`. For example. `StageAward:Reverse()["StageAward_FullComboW3"]` will return the corresponding enum value. See [documentation](https://itgmania.github.io/lua-for-itgmania/LuaAPI#Namespaces-Enum) for details.

I definitely see StageAwards in a `ECFA-stats.xml`. See: https://pastebin.com/rpHcxTxq . You won't get a `StageAward` for a great full combo because that falls into the `W4` window in FA+ mode. That's the only manual thing we will need to compute.

Reply:
Shoot, I thought I had tested with a pfc 🤦. Looks like only my Outfox test had the PFC, which always would return FullComboW3 anyway because of that bug.

Good catch. I've changed the implementation. I'll resolve this conversation for now, but if you've got any new comments/concerns then let em rip.

Another note, apparently Outfox is unable to recognize the Enum:Compare function... Was trying my best to maintain compatibility as much as possible, but this implementation now wholly won't work with Outfox. Not a big deal though since I've got a commit that works with counting a steps manually, so it could be cherrypicked into that fork should they so choose.